components/places/src/storage.rs

```diff
@@ -263,21 +270,314 @@ fn add_visit(db: &impl ConnExt,
+        pub visit_type: Option<VisitTransition>,
+    }
 
-    db.execute_named("
-        UPDATE moz_places
-        SET frecency = :frecency
-        WHERE id = :page_id",
-        &[(":frecency", &score), (":page_id", &id.0)])?;
+    impl FetchedVisit {
+        pub fn from_row(row: &Row) -> Result<Self> {
+            Ok(Self {
+                is_local: row.get_checked("is_local")?,
+                visit_date: row.get_checked::<_, Option<Timestamp>>("visit_date")?
+                     .unwrap_or_default(),
+                visit_type: VisitTransition::from_primitive(
+                                row.get_checked::<_, Option<u8>>("visit_type")?.unwrap_or(0)
+                            ),
+            })
+        }
+    }
 
-    Ok(())
-}
+    #[derive(Debug)]
+    pub struct FetchedVisitPage {
+        pub url: Url,
+        pub guid: SyncGuid,
+        pub row_id: RowId,
+        pub title: String,
+    }
+
+    impl FetchedVisitPage {
+        pub fn from_row(row: &Row) -> Result<Self> {
+            Ok(Self {
+                url: Url::parse(&row.get_checked::<_, Option<String>>("url")?.expect("non null column"))?,
+                guid: SyncGuid(row.get_checked::<_, Option<String>>("guid")?.expect("non null column")),
+                row_id: row.get_checked("id")?,
+                title: row.get_checked::<_, Option<String>>("title")?.unwrap_or_default(),
+            })
+        }
+    }
+
+    pub fn fetch_visits(db: &Connection, url: &Url, limit: usize) -> Result<Option<(FetchedVisitPage, Vec<FetchedVisit>)>> {
+        // We do this in 2 steps - "do we have a page" then "get visits"
+        let page_sql = "
+          SELECT guid, url, id, title
+          FROM moz_places h
+          WHERE url_hash = hash(:url) AND url = :url";
+
+        let page_info = match db.try_query_row(page_sql, &[(":url", &url.clone().into_string())], FetchedVisitPage::from_row, true)? {
+            None => return Ok(None),
+            Some(pi) => pi,
+        };
+
+        let mut stmt = db.prepare("
+          SELECT is_local, visit_type, visit_date
+          FROM moz_historyvisits
+          WHERE place_id = :place_id
+          LIMIT :limit")?;
+        let visits = stmt.query_and_then_named(&[(":place_id", &page_info.row_id),
+                                                 (":limit", &(limit as u32))],
+                                               FetchedVisit::from_row)?
+                         .collect::<Result<Vec<_>>>()?;
+        Ok(Some((page_info, visits)))
+    }
+
+    /// Apply history visit from sync. This assumes they have all been
+    /// validated, deduped, etc - it's just the storage we do here.
+    pub fn apply_synced_visits(db: &Connection,
+                               new_guid: &SyncGuid,
+                               existing_guid: &Option<SyncGuid>,
+                               url: &Url, title: &Option<String>,
+                               visits: &Vec<HistoryRecordVisit>) -> Result<()> {
```

Comment:
Nit: You basically always should pass `&[T]` over `&Vec<T>`. And the conversion between them is automatic, so there's no need to change elsewhere.